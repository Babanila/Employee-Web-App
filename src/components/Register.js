import React, { useState } from "react";
import { Redirect } from "react-router-dom";
import axios from "axios";
import Header from "./Header";
import ErrorComp from "./ErrorComp";
import { RegisterStyles } from "../styles/RegisterStyles";

const Register = () => {
  const [id, setId] = useState("");
  const [employee_name, setName] = useState("");
  const [employee_age, setAge] = useState("");
  const [employee_salary, setSalary] = useState("");
  const [profile_image, setPicture] = useState("");
  const [error, setError] = useState({ message: "" });
  const [redirect, setRedirect] = useState(false);

  // Create new employee
  const createEmployeeData = async data => {
    const url = "http://dummy.restapiexample.com/api/v1/create";
    try {
      const res = await axios.post(url, data);

      // Get the Id of the created employee from the response
      setId(res.data.id);
      setRedirect(true);
    } catch (e) {
      setError(e);
    }
  };

  const handleSubmit = event => {
    event.preventDefault();
    createEmployeeData({
      name: employee_name,
      age: employee_age,
      salary: employee_salary,
      picture: profile_image
    });
  };

  return (
    <RegisterStyles>
      <Header pageTitle="Register Page" />
      {error.message.length !== 0 ? (
        <ErrorComp data={error} className="register-error" />
      ) : redirect ? (
        <Redirect to={`/`} />
      ) : (
        <form className="register-form" onSubmit={handleSubmit}>
          <h1>New Employee Registration Form</h1>
          <h3>Kindly fill all the fields below.</h3>
          <input
            className="form-input"
            name="employee_name"
            value={employee_name}
            placeholder="Name"
            onChange={e => {
              setName(e.target.value);
            }}
          />
          <input
            className="form-input"
            name="employee_age"
            value={employee_age}
            placeholder="Age"
            onChange={e => {
              setAge(e.target.value);
            }}
          />
          <input
            className="form-input"
            name="employee_salary"
            value={employee_salary}
            placeholder="Salary"
            onChange={e => {
              setSalary(e.target.value);
            }}
          />
          <input
            className="form-input"
            name="profile_image"
            value={profile_image}
            placeholder="Picture"
            onChange={e => {
              setPicture(e.target.value);
            }}
          />
          <button className="form-button" type="submit">
            Register
          </button>
        </form>
      )}
    </RegisterStyles>
  );
};

export default Register;
