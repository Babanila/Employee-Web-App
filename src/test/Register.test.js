import React from "react";
import renderer from "react-test-renderer";
import { render, cleanup, fireEvent, waitForElement } from "@testing-library/react";
import { shallow } from "enzyme";
import Register from "../components/Register";

describe("Register Component", () => {
  afterEach(cleanup);

  it("should renders without crashing", () => {
    const RegisterComponent = renderer.create(<Register />).toJSON();
    expect(RegisterComponent).toMatchSnapshot();
  });

  it("should render correct number of element in the component", () => {
    const RegisterWrapper = shallow(<Register />);
    expect(RegisterWrapper.find("form").exists).toBeTruthy();
    expect(RegisterWrapper.find("h1").exists).toBeTruthy();
    expect(RegisterWrapper.find("h3").exists).toBeTruthy();
    expect(RegisterWrapper.find("input").exists).toBeTruthy();
    expect(RegisterWrapper.find("button").exists).toBeTruthy();
    expect(RegisterWrapper.find("form").length).toEqual(1);
    expect(RegisterWrapper.find("h1").length).toEqual(1);
    expect(RegisterWrapper.find("h3").length).toEqual(1);
    expect(RegisterWrapper.find("input").length).toEqual(4);
    expect(RegisterWrapper.find("button").length).toEqual(1);
  });

  it("should renders correct employee components", async () => {
    const { getByPlaceholderText } = render(<Register />);

    expect(getByPlaceholderText("Name")).toHaveTextContent("");
    expect(getByPlaceholderText("Age")).toHaveTextContent("");
    expect(getByPlaceholderText("Salary")).toHaveTextContent("");
    expect(getByPlaceholderText("Picture")).toHaveTextContent("");
  });

  it("should renders employee details  correctly", async () => {
    const { getByPlaceholderText, getByText, queryByText, queryByTestId } = render(<Register />);

    expect(getByPlaceholderText("Name")).toHaveTextContent("");
    expect(getByPlaceholderText("Age")).toHaveTextContent("");
    expect(getByPlaceholderText("Salary")).toHaveTextContent("");
    expect(getByPlaceholderText("Picture")).toHaveTextContent("");
  });

  xit("should post to API and displays the registered user", async () => {
    const fakeAxios = {
      post: jest.fn(() =>
        Promise.resolve({
          data: {
            id: 34,
            name: "John",
            age: 25,
            salary: 4000,
            picture: "john.jpg"
          }
        })
      )
    };
    const url = "http://dummy.restapiexample.com/api/v1/create";
    const { getByText, getByTestId } = render(<Register />);

    fireEvent.click(getByTestId("register-button"));
  });
});
