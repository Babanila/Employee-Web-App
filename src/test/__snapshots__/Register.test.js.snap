// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Register Component should renders without crashing 1`] = `
<div
  className="css-vs0298"
>
  <div
    className="css-yhx7tk"
  >
    <button
      className="toggle-button css-9glsrg"
      onClick={[Function]}
    >
      <div
        className="button-line"
      />
      <div
        className="button-line"
      />
      <div
        className="button-line"
      />
    </button>
    <h1>
      Registeration Page
    </h1>
    <div
      className="spacer"
    />
  </div>
  <form
    className="register-form"
    onSubmit={[Function]}
  >
    <h1>
      New Employee Registration Form
    </h1>
    <h3>
      Kindly fill all the fields below.
    </h3>
    <input
      className="form-input"
      data-testid="employee_name"
      name="employee_name"
      onChange={[Function]}
      placeholder="Name"
      value=""
    />
    <input
      className="form-input"
      data-testid="employee_age"
      name="employee_age"
      onChange={[Function]}
      placeholder="Age"
      value=""
    />
    <input
      className="form-input"
      data-testid="employee_salary"
      name="employee_salary"
      onChange={[Function]}
      placeholder="Salary"
      value=""
    />
    <input
      className="form-input"
      data-testid="picture"
      name="profile_image"
      onChange={[Function]}
      placeholder="Picture"
      value=""
    />
    <button
      className="form-button"
      data-testid="register-button"
      type="submit"
    >
      Register
    </button>
  </form>
</div>
`;
